File: script/definitions.js
let DEBUG = false; // Show debug info in the console
let BENCHMARK_TICKS = 100; // Number of simulation steps to run the benchmark for
let BENCHMARK_BATCHES = 10; // Benchmarks are averaged over this many runs
let BENCHMARK_DENSITY = 0.05; // Proportion of tiles that are not air in benchmark worlds, this ensures tiles can move around

const TPS = 30;
const START_PAUSED = true;

const TILESET = {
  AIR: "skyblue",
  SOIL: "peru",
  SAND: "sandybrown",
  STONE: "slategray",
  WORKER: "red",
  QUEEN: "blueviolet",
  EGG: "white",
  CORPSE: "black",
  PLANT: "olivedrab",
  WATER: "blue",
  FUNGUS: "teal",
  PEST: "fuchsia",
  TRAIL: "yellow",
};

// Tiles that can be overridden with the brush
const PAINTABLE_MASK = ["AIR", "SOIL", "SAND"];

const RAIN_FREQ = 2500; // How often (in game ticks) it rains
const RAIN_TIME = 500; // How long (in game ticks) it rains for
const PEST_FREQ = 100; // How often (in game ticks) a pest enters the map
const PEST_START = 4000; // How long (in game ticks) before pests can spawn

// These are modifiable during run-time using the browser console but only take effect after the game is reset
let ROW_COUNT = 100; // World height
let COL_COUNT = 100; // World width

// These are modifiable during run-time using the browser console and take effect immediately
let CHUNK_SIZE = 20; // Optimisation for faster searching, must be factor of ROW_COUNT and COL_COUNT
let KILL_PROB = 0.01; // Chance per tick for each hazard to kill
let EVAPORATE_PROB = 0.008; // Chance per tick for water to evaporate
let CONVERT_PROB = 0.03; // Chance per tick for fungus/plant to convert
let GROW_PROB = 0.1; // Base chance per tick for a lone plant tile to grow (reduced by crowding)
let EGG_LAY_PROB = 0.5; // Chance that an egg is laid when a queen eats fungus
let EGG_HATCH_PROB = 0.005; // Chance per tick for an egg to hatch
let EGG_QUEEN_PROB = 0.01; // Chance for a hatching egg to be a queen
let QUEEN_SPEED = 0.1; // The queen will only act this proportion of ticks
let QUEEN_RANGE = 20; // Search radius of queens
let QUEEN_FUNGUS_MIN = 8; // Minimum fungus tiles in range for queen to feed
let PEST_SEEK_PROB = 0.2; // Chance per tick for a pest to seek a target
let PEST_RANGE = 20; // Search radius of pests
let WORKER_RANGE = 15; // Search radius of workers

// Tiles ants can climb
const CLIMB_MASK = [
  "SOIL",
  "SAND",
  "STONE",
  "FUNGUS",
  "CORPSE",
  "WORKER",
  "EGG",
];

// Tiles ants can move through (also used by pests when hunting)
const WALK_MASK = ["AIR", "CORPSE", "EGG", "PLANT", "FUNGUS", "TRAIL"];

// Tiles pests can move through when wandering
const ROAM_MASK = ["AIR", "CORPSE", "FUNGUS", "TRAIL"];

// Tiles plants can grow into
const PLANT_GROW_MASK = ["AIR", "WATER", "CORPSE", "TRAIL"];

// Tiles pests will seek to kill
const PEST_TARGET_MASK = ["EGG", "QUEEN", "WORKER"];

// Tiles water will kill
const WATER_KILL_MASK = ["WORKER", "QUEEN", "EGG", "PEST"];

// Tiles workers can push when they move
const PUSH_MASK = ["PLANT", "CORPSE"];

// module exports all variables
if (typeof module === 'object') {
  module.exports = {
    DEBUG,
    BENCHMARK_TICKS,
    BENCHMARK_BATCHES,
    BENCHMARK_DENSITY,
    TPS,
    START_PAUSED,
    TILESET,
    PAINTABLE_MASK,
    ROW_COUNT,
    COL_COUNT,
    CHUNK_SIZE,
    RAIN_FREQ,
    RAIN_TIME,
    PEST_FREQ,
    PEST_START,
    KILL_PROB,
    EVAPORATE_PROB,
    CONVERT_PROB,
    GROW_PROB,
    EGG_LAY_PROB,
    EGG_HATCH_PROB,
    EGG_QUEEN_PROB,
    QUEEN_SPEED,
    QUEEN_RANGE,
    QUEEN_FUNGUS_MIN,
    PEST_SEEK_PROB,
    PEST_RANGE,
    WORKER_RANGE,
    CLIMB_MASK,
    WALK_MASK,
    ROAM_MASK,
    PLANT_GROW_MASK,
    PEST_TARGET_MASK,
    WATER_KILL_MASK,
    PUSH_MASK
  };
}

File: src/definitions.rs
#[derive(Copy, Clone, PartialEq)]
pub enum TileSet {
    AIR,
    SOIL,
    SAND,
    STONE,
    WORKER,
    QUEEN,
    EGG,
    CORPSE,
    PLANT,
    WATER,
    FUNGUS,
    PEST,
    TRAIL,
}

impl TileSet {
    fn color(&self) -> &'static str {
        match self {
            TileSet::AIR => "skyblue",
            TileSet::SOIL => "peru",
            TileSet::SAND => "sandybrown",
            TileSet::STONE => "slategray",
            TileSet::WORKER => "red",
            TileSet::QUEEN => "blueviolet",
            TileSet::EGG => "white",
            TileSet::CORPSE => "black",
            TileSet::PLANT => "olivedrab",
            TileSet::WATER => "blue",
            TileSet::FUNGUS => "teal",
            TileSet::PEST => "fuchsia",
            TileSet::TRAIL => "yellow",
        }
    }
}

#[derive(Clone)]
pub struct Chunk {
    pub air: i32,
    pub soil: i32,
    pub sand: i32,
    pub stone: i32,
    pub worker: i32,
    pub queen: i32,
    pub egg: i32,
    pub corpse: i32,
    pub plant: i32,
    pub water: i32,
    pub fungus: i32,
    pub pest: i32,
    pub trail: i32,
}

impl Chunk {
    pub fn get(&self, tile: &TileSet) -> i32 {
        match tile {
            TileSet::AIR => self.air,
            TileSet::SOIL => self.soil,
            TileSet::SAND => self.sand,
            TileSet::STONE => self.stone,
            TileSet::WORKER => self.worker,
            TileSet::QUEEN => self.queen,
            TileSet::EGG => self.egg,
            TileSet::CORPSE => self.corpse,
            TileSet::PLANT => self.plant,
            TileSet::WATER => self.water,
            TileSet::FUNGUS => self.fungus,
            TileSet::PEST => self.pest,
            TileSet::TRAIL => self.trail,
        }
    }
}

#[derive(Clone)]
pub struct Definitions {
    pub debug: bool,
    pub benchmark_ticks: i32,
    pub benchmark_batches: i32,
    pub benchmark_density: f64,
    pub tps: i32,
    pub start_paused: bool,
    pub row_count: i32,
    pub col_count: i32,
    pub chunk_size: i32,
    pub rain_freq: i32,
    pub rain_time: i32,
    pub pest_freq: i32,
    pub pest_start: i32,
    pub kill_prob: f64,
    pub evaporate_prob: f64,
    pub convert_prob: f64,
    pub grow_prob: f64,
    pub egg_lay_prob: f64,
    pub egg_hatch_prob: f64,
    pub egg_queen_prob: f64,
    pub queen_speed: f64,
    pub queen_range: i32,
    pub queen_fungus_min: i32,
    pub pest_seek_prob: f64,
    pub pest_range: i32,
    pub worker_range: i32,
    pub paintable_mask: [TileSet; 3],
    pub climb_mask: [TileSet; 7],
    pub walk_mask: [TileSet; 6],
    pub roam_mask: [TileSet; 4],
    pub plant_grow_mask: [TileSet; 4],
    pub pest_target_mask: [TileSet; 3],
    pub water_kill_mask: [TileSet; 4],
    pub push_mask: [TileSet; 2],
}

impl Definitions {
    pub fn new() -> Self {
        Definitions {
            debug: true,
            benchmark_ticks: 100,
            benchmark_batches: 10,
            benchmark_density: 0.05,
            tps: 30,
            start_paused: true,
            row_count: 100,
            col_count: 100,
            chunk_size: 20,
            rain_freq: 2500,
            rain_time: 500,
            pest_freq: 100,
            pest_start: 4000,
            kill_prob: 0.01,
            evaporate_prob: 0.008,
            convert_prob: 0.03,
            grow_prob: 0.1,
            egg_lay_prob: 0.5,
            egg_hatch_prob: 0.005,
            egg_queen_prob: 0.01,
            queen_speed: 0.1,
            queen_range: 20,
            queen_fungus_min: 8,
            pest_seek_prob: 0.2,
            pest_range: 20,
            worker_range: 15,
            paintable_mask: [TileSet::AIR, TileSet::SOIL, TileSet::SAND],
            climb_mask: [
                TileSet::SOIL,
                TileSet::SAND,
                TileSet::STONE,
                TileSet::FUNGUS,
                TileSet::CORPSE,
                TileSet::WORKER,
                TileSet::EGG,
            ],
            walk_mask: [
                TileSet::AIR,
                TileSet::CORPSE,
                TileSet::EGG,
                TileSet::PLANT,
                TileSet::FUNGUS,
                TileSet::TRAIL,
            ],
            roam_mask: [TileSet::AIR, TileSet::CORPSE, TileSet::FUNGUS, TileSet::TRAIL],
            plant_grow_mask: [
                TileSet::AIR,
                TileSet::WATER,
                TileSet::CORPSE,
                TileSet::TRAIL,
            ],
            pest_target_mask: [TileSet::EGG, TileSet::QUEEN, TileSet::WORKER],
            water_kill_mask: [
                TileSet::WORKER,
                TileSet::QUEEN,
                TileSet::EGG,
                TileSet::PEST,
            ],
            push_mask: [TileSet::PLANT, TileSet::CORPSE],
        }
    }
}
